text/0000-assign-param-syntax.md

```diff
@@ -0,0 +1,115 @@
+
+```pony
+class val UseFFIDecl is (AST & UseDecl)
+  let attachments: (Attachments | None)
+  let name: (Id | LitString)
+  let return_type: TypeArgs
+  let params: Params
+  let partial: (Question | None)
+  let guard: (IfDefCond | None)
+
+  new val create(name: String, @return_type, @params, @attachments = None) =>
+    @name = Id(name)
+    @partial = None
+    @guard = None
+```
+
+As you can see, this syntax opens new opportunities but shouldn't preclude any old ones. Thus, it is not a breaking change or required learning, and it merely offers convenience for those looking for it. There are no new keywords or symbols needed in the parser.
+
+For many more syntax examples, see [the `.mare` files in the Mare compiler's codebase](https://github.com/jemc/mare/tree/master/src/prelude), where this syntax already exists and is parsed by the parser, along with the other syntax changes included there. This RFC and others to come are attempting to try to get some general consensus agreement on the desireability of at least some of the syntax changes that have been introduced in that work.
+
+## Compiler implementation discussion
+
+Internally, the names of these parameters are anonymous/hygienic ids, meaning that they cannot be used by any explicit code and will not collide with any such names - they are are anonymous references that are immediately assigned at the beginning of the function body.
+
+Note that they must also be `consume`d in the assignment, so that the assignment will work properly for `iso`- and `trn`-typed fields. Again, this is an internal implementation detail just to make it behave as the user expects.
+
+One additional implementation wrinkle is how to handle the case of non-sendable field caps being used as assign params to functions/constructors that allow only sendable parameters. For example, the case of an `ref` field  on an assign param of an actor's constructor. In this case, the assign param cannot be the same cap as the field as it usually would be, since `ref` parameters are not sendable to the actor's constructor. Instead we must "lift" the cap of the assign param to the "nearest" sendable cap that can be assigned to that field cap - in this case, it would be `iso`. This follows the principle of least surprise because this is exactly what the programmer would have to do explicitly to make such a code pattern work properly with capabilities.
+
+# How We Teach This
+
+Add a new section to the tutorial to teach this.
+
+Change some of the standard library to demonstrate it in action.
+
+# How We Test This
+
+New compiler tests that demonstrate this case, including the edge cases mentioned in the implementation details above.
+
+# Drawbacks
+
+* New patterns to learn (though no codebases will be required to use these patterns, you have to learn the new patterns if you want to write or read code that uses them).
+* Added compiler complexity.
+
```

Comment:
Symbols/glyphs are hard to understand and search for, so this makes code less explicit and thus possibly harder to learn, or more intimidating

Reply:
I personally don't really buy this argument in this case, because if we used the `@` symbol for `this.` (and if `this.` were mandatory), as proposed in the other RFC, this symbol would be used in nearly every function of every file of every Pony program.

Sure, somebody who was learning the language for the first time won't immediately know what it means, but they'd learn what it means within the first 5-10 minutes of learning, and they wouldn't forget because it would be so ubiquitous. Searchability of the symbol in search engines is not really relevant, because anyone encountering it for the first time would also be learning the language for the first time, so they'd already be looking at the tutorial.

I actually think that this argument is a much more compelling reason to *not* use `@` for FFI or bare functions, and to use it for the much more ubiquitous `this.` instead. FFI code is much more obscure, and a user could be using Pony for years before encoutnering their first reading of FFI code. Hence, if and when they do encounter the `@` symbol for FFI or a bare function they are likely to be confused, and find it hard to search for.

Reply:
This now has `this.` instead of `@` so this is resolved for now.